contracts/vanchor/src/contract.rs

```diff
@@ -0,0 +1,528 @@
+            ext_data_args.extend_from_slice(&fee_bytes);
+            ext_data_args.extend_from_slice(&ext_data.encrypted_output1);
+            ext_data_args.extend_from_slice(&ext_data.encrypted_output2);
+
+            let computed_ext_data_hash =
+                Keccak256::hash(&ext_data_args).map_err(|_| ContractError::HashError)?;
+            if computed_ext_data_hash != proof_data.ext_data_hash {
+                return Err(ContractError::Std(StdError::GenericErr {
+                    msg: "Invalid ext data".to_string(),
+                }));
+            }
+
+            let abs_ext_amt = ext_amt.unsigned_abs();
+            // Making sure that public amount and fee are correct
+            if Uint256::from(ext_data_fee) > vanchor.max_fee {
+                return Err(ContractError::Std(StdError::GenericErr {
+                    msg: "Invalid fee amount".to_string(),
+                }));
+            }
+
+            if Uint256::from(abs_ext_amt) > vanchor.max_ext_amt {
+                return Err(ContractError::Std(StdError::GenericErr {
+                    msg: "Invalid ext amount".to_string(),
+                }));
+            }
+
+            // Public amounnt can also be negative, in which
+            // case it would wrap around the field, so we should check if FIELD_SIZE -
+            // public_amount == proof_data.public_amount, in case of a negative ext_amount
+            let calc_public_amt = ext_amt - ext_data_fee as i128;
+            let calc_public_amt_bytes =
+                element_encoder(&ArkworksIntoFieldBn254::into_field(calc_public_amt));
+            if calc_public_amt_bytes != proof_data.public_amount {
+                return Err(ContractError::Std(StdError::GenericErr {
+                    msg: "Invalid public amount".to_string(),
+                }));
+            }
+
+            // Construct public inputs
+            let chain_id_type_bytes = element_encoder(
+                &compute_chain_id_type(vanchor.chain_id, &COSMOS_CHAIN_TYPE).to_le_bytes(),
+            );
+
+            let mut bytes = Vec::new();
+            bytes.extend_from_slice(&proof_data.public_amount);
+            bytes.extend_from_slice(&proof_data.ext_data_hash);
+            for null in &proof_data.input_nullifiers {
+                bytes.extend_from_slice(null);
+            }
+            for comm in &proof_data.output_commitments {
+                bytes.extend_from_slice(comm);
+            }
+            bytes.extend_from_slice(&element_encoder(&chain_id_type_bytes));
+            for root in &proof_data.roots {
+                bytes.extend_from_slice(root);
+            }
+
+            let verifier = VANCHORVERIFIER.load(deps.storage)?;
+            let result = match (
+                proof_data.input_nullifiers.len(),
+                proof_data.output_commitments.len(),
+            ) {
+                (2, 2) => verify(verifier, bytes, proof_data.proof)?,
```

Comment:
Can we add (16, 2) as well and task out integrating the 16-2 join/split tx. CC @dutterbutter to triage task assignment:
- Add 2 verifiers to the contract (small / large)
- Generate trusted parameters for 16-2 join/split (we should add ones for 2 vanchor bridge, 8 vanchor bridge, and 32 vanchor bridge)
- Add them to protocol-substrate-fixtures
- Hook them up here and add integration tests.